### PR TITLE
Add UML to settle down the project's structure

### PR DESCRIPTION
**Pull request by starsasumi**
#28

OK to merge.

**Note**: To preview the UML (SVG file), click the "View" button in the "Files changed" tab on this PR. And you can right-click the image to open it in a new navigator tab to watch a larger image.
screen shot 2015-10-26 at 11 25 42

What I did:
1. Add initial UML that reflect our existed codes
2. Move SVGInterface to model package
3. Adapt factory pattern in the creation of shape
4. Add interfaces for script language
5. Remove all presentation related apis out of shapes
6. Add interaction mechanism between presentations and shapes
7. Add demo codes in `src/model/presentation/`

### CODE REVIEW

**Comment by Thomas-dot-G** [comment]
Then assign yourself, it means you are still working on it

**Comment by Thomas-dot-G** [comment]
Model.Shape and Model.ShapeFactory are a good idea that we should implement in the actual code.
That may need a little more design tho^^
(The goal would be to be able to reproduce that with the project^^)

**Comment by starsasumi** [comment]
@Thomas-dot-G @anqitong 

After thinking in our model, I think we should move all xml related codes out of our shapes because: 

> Au final, au moins deux interprétations doivent être réalisées. L'une des interprétations doit correspondre à la production d'un fichier au format SVG.

So, all shapes would only keep its geometric properties. Then we have to create an interface to encapsulate presentation related algos which its concrete subclasses would handle things like xml generations. In the end, we should use some design pattern to make shapes and presentations work together without knowing too much about others.

For the design patterns, I am thinking about using the [Strategy Pattern](https://fr.wikipedia.org/wiki/Stratégie_%28patron_de_conception%29) to alter between 2 types of presentations and using the [State Pattern](https://fr.wikipedia.org/wiki/Wikipédia:Accueil_principal) to generation presentations for different shapes. 

Please tell me what you think about this?

**Comment by Thomas-dot-G** [comment]
Yeah, why not, but I am not entirely sure what you mean in termes of interface
Beacause in the shapes (specific ones) one method is to get the xml, but another can be for something else (little bit like a visitor)

**Comment by starsasumi** [comment]
@Thomas-dot-G but if like we need to add a third presentation, using methods like getXML() would cause us to open all specific shapes' codes add add method.

What I mean is like this:

```
interface Shape {
    ShapeName getShapeName();
}

interface Presentation {
    void addShapes(List<Shape> list);
    void display()
}

interface ShapeState {
    void setShape(Shape shape);
    String getDrawing();
}
```

Now we delegate real algos to concrete ShapeState objects. So when we add shapes or add presentations, we don't have to modify much. Instead, we add an implementation of ShapeState.

**Comment by starsasumi** [comment]
I have update the UML to describe the interaction mechanism between presentations and shapes. Now it means that we add shapes to a presentation (like XML), and a presentation loops through these shapes and new a related ShapeState object to generate the real presentation.

**Comment by Thomas-dot-G** [comment]
Thanks, I understand better.
Yeah, you are right, but that means that in a class, their will be the xml of every shape (only xml) ?

**Comment by starsasumi** [comment]
In my mind, a XML (implements Presentation) contains a list of shape objects. Whenever its `display()` or other similar methods (like `writeXML()`) get called, XML re-generates a real xml file from the list by calling `getDrawing()` of all shapes.

**Comment by Thomas-dot-G** [comment]
yes, but getDrawing is not specific and not xml.
The xml generated depends of the object being called, that is why for now the xml is in each shape.
With the interface there, we'll need a class with the xml of each shape thanks to the object (but no need for getDrawing() with, as general, will only get the object itself.

**Comment by starsasumi** [comment]
We could define one `ShapeState` for one pair of `Shape` - `Presentation`, like

```
class CircleXML implements ShapeState {
    Circle circle
    Object getDrawing() { return "<circle ..." + circle.xyz + " ... >"; }
}
```

And all shapes that implement `Shape` keep know nothing about how to draw itself, they store nothing but geometric properties like point & radius. The `ShapeState` of one shape should know well about these properties.

In that case, no matter we add shapes or presentations, we add subclasses of ShapeState, and nothing but `enum ShapeName` would need to be modified (["be open for extension, but closed for modification"](https://en.wikipedia.org/wiki/Open/closed_principle)) .

For the return type of `getDrawing()`, we can change it to `Object` because only one type of presentation would use that shapeState.

**Comment by Thomas-dot-G** [comment]
Ok, just creating more class, that is a way, yes.
Still not sure about the getDrawing however. What is its purpose ?
It also means that in the main, we need to create a circle and a CircleXML ? Wouldn't it be a little too heavy.

I have an idea of how (maybe its your's that I didn't understand):
Transform the getXML() into getDrawing(), and the getDrawing, and in the ShapeType, their is a var interpretation, that could be xml, then, the var actually points to an other type (ex CirclePresentation) which has every way of presenting the circle. If the var is xml, the function will get the xml...

**Comment by starsasumi** [comment]
Yep it's a little bit heavy. The purpose is just to loose coupling. And because Presentation knows about how to draw, so Presentation should know what a ShapeState return in `getDrawing()`. For example, SVGFile knows about xml, and it knows `Object getDrawing()` in `CircleXML implements ShapeState` returns String.

To do this loose coupling or not depends on your opinions. Indeed, a `getXml()` and a `getOtherPresentation()` in every shape is an instinctive approach that works. If I hack this project alone I would probably do that too. Loosing couplings is just to show we are using some architecture tricks 😝

For your new purpose, I think it is equivalent to having getXml() and getOther() in shapes.

**Comment by Thomas-dot-G** [comment]
I agree with you.
Could you just describe to me (briefly) how it would be done in the code.
Because I'm not sure how you would implement that "getDrawing" thing.
I think I will understand. you can just create an object presentation, and if you give it your shape, it will automatically create the right shape state, so the getDrawing would be in the master interface.
Is that it ?

**Comment by anqitong** [comment]
Anyway we should not be creating both a Circle and a CircleXML. From the user side, it is not convenient. 
About the type of getDrawing, since we know it is String, I think the return type should also stay as String

**Comment by starsasumi** [comment]
For @anqitong , user know nothing about this. Users just select which presentation they want.

@Thomas-dot-G I've added demo codes in src/model/presentation

Since only task-delegating codes are added, performance should not be affect by this model (UI codes >>> IO codes >>> pure logic in memory). The session 1's live demo in the cours does something similar.

**Comment by Thomas-dot-G** [comment]
Ok, but that code does not help that much^^ sorry.

I thing that the fabric should do the trick:
Ex, you create a Circle with a fabric an the name of the presentation, and the fabric create a CircleXML (with a Circle).
So the user won't see the difference.
In that case, we can use what you described. And each ShapeState will have the same methods.

Agree ?

**Comment by starsasumi** [comment]
It seems good to me 👍

**Comment by Thomas-dot-G** [comment]
Great,
I suggest we merge the current PR, then try to reorganize all that.
Maybe just change the UML for the fabric, I that's all.
Not sure that that code will be used exactly like that. We should rather change the current one.

**Comment by starsasumi** [comment]
OK, I clean up a little bit the UML.

**Comment by anqitong** [comment]
Agree for the merge

**Comment by starsasumi** [comment]
I update the UML to fit demo codes. It should be ok to merge now.

**Comment by Thomas-dot-G** [comment]
There is still some code that already exist

Why changed in the UML?, I don't see changes

**Comment by starsasumi** [comment]
Got it, now I keep only two interfaces that should be in `model.presentation`

**Comment by Thomas-dot-G** [comment]
Yeah, but shapesFactory also create the presentationState shape. That the only thing missing for me
(And the code to remove in that PR)

File: src/model/presentation/Presentation.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import model.ShapesName;
+
+import java.util.List;
+
+/**
+ * Created by Xiaxing SHI on 27/10/15.
+ */
+public interface Presentation {
+    void addShapes(List<Shape> shapes);
```
Comment (Thomas-dot-G) [inline src/model/presentation/Presentation.java]: Add javadoc with the fact that addShapes creates the specific shapeState according to the type of shapes given

**Comment by starsasumi** [comment]
I've update comments to explicit the model

**Comment by Thomas-dot-G** [comment]
No, the javadoc is in the code, the lines where I commented above, /*\* created by, is not really explicit **/

**Comment by starsasumi** [comment]
Javadoc added 😁. Sorry for assigning you before I finished all modifications.

File: src/model/presentation/Presentation.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    void display();
+
+    /**
+     * This method is a tool to create concrete ShapeStates according to the type of shapes given. A typical implementation is to
+     * <code>switch (shape.getName())</code>
+     * @param shape the shape to render by ShapeState
+     * @return a concrete ShapeState object that can render shape
+     */
+    ShapeState getShapeState(ShapesType shape);
```
Comment (Thomas-dot-G) [inline src/model/presentation/Presentation.java:31]: create or just get ?

**Comment by starsasumi** [comment]
Both. One new object a time or only one object for a type, all two ways work.  And as an interface, it don't care about this kind of implementation detail.

**Comment by Thomas-dot-G** [comment]
it was just about the name that was not really matching the javadoc.
But in that case, for me, ready to merge.
(Btw, you have other things assigned)

**Pull request merged**